### PR TITLE
[Feature][Model] Support DeepSeek V4 Flash Vision on Ascend

### PR DESCRIPTION
**Pull request by GDzhu01**
### What this PR does / why we need it

Adds `DeepSeek-V4-Flash-Vision-Exp` support to vLLM Ascend `main`, paired with
the vLLM v0.27.x API line.

Roadmap and remaining validation tracking: #15462

Upstream behavior reference: vllm-project/vllm#54566

The paired vLLM release does not contain the DeepSeek-V4 vision processor and
vision tower from the upstream PR. This change therefore carries the required
v0.27-compatible implementation in vLLM Ascend without importing unmerged
vLLM modules.

### Implementation

- registers `DeepseekV4ForConditionalGeneration` and routes vision
  checkpoints to an Ascend multimodal wrapper;
- provides the image processor, ViT, aligner, sentinel embeddings, placeholder
  expansion, and multimodal weight mapping required by vLLM v0.27;
- preserves raw input token IDs through prefill for DeepSeek-V4 MoE routing;
- implements hybrid `bias_vl` routing: text tokens retain hash/correction-bias
  routing while image sentinel tokens use dynamic routing with `bias_vl`;
- builds paged original-KV sparse indices that make each complete image span
  bidirectionally visible while keeping causal SWA outside image spans;
- passes original-KV image indices and compressed-KV top-k indices to the
  Ascend sparse-attention operator independently and simultaneously;
- supports ModelSlim W8A8 checkpoints whose language quantization description
  remains in the causal-LM `model.*` / `lm_head.*` namespace;
- keeps config-convertor registration safe in spawned DP engine processes.
- enables DeepSeek-V4 Vision with a text-only DSpark drafter: image inputs are
  encoded by the target during prefill, while DSpark proposes subsequent text
  decode tokens; restores the draft architecture after VL config conversion,
  exposes the EAGLE3 hidden-state interface, and ignores draft-only vision
  router bias tensors without changing text routing.

### Runtime validation completed

Validated in a 16-NPU Ascend container with:

- vLLM: `6e448d0ea9bf3d88d898b65449ca6dc2aec170ac` (v0.27.1 API line)
- checkpoint: DeepSeek-V4-Flash-Vision-Exp W8A8, with all BF16 `bias_vl`
  tensors included in the checkpoint overlay
- release topology: TP4 / DP4 / EP16, block size 32

Results:

- targeted processor, ModelSlim, MoE router, and DSA tests pass (90 tests in
  the combined suite; 59 tests in the focused follow-up suite);
- the Ascend sparse-attention operator accepts widened original indices and
  simultaneous original/compressed indices with finite BF16 output;
- all 79 checkpoint shards load successfully with the multimodal wrapper;
- eager OpenAI-compatible serving passes text-only, single-image, ordered
  two-image, deterministic repeated-image, and invalid-image cases;
- the single-image OCR result correctly reads `Hello, AI world!`, and the
  ordered two-image result preserves both image order and text;
- ACLGraph mode passes torch compilation, PIECEWISE mixed prefill/decode
  capture, FULL decode capture, and runtime graph replay on all four DP
  leaders;
- greedy text and single-image outputs are identical between eager and graph
  modes;
- invalid image data returns an explicit HTTP 400 response.
- DSpark eager and FULL_DECODE_ONLY target-graph serving pass text, single-image,
  two-image, and 149-token long-decode cases with byte-identical outputs;
- eager and graph runs each accepted 153 of 180 draft tokens (85.0%); four
  concurrent graph-mode long decodes all passed, reaching 656/750 cumulative
  accepted/draft tokens (87.47%).
- three 149-token requests pinned concurrently to one DP rank also pass
  byte-identically inside the captured 24-token graph range, with 375/414
  accepted/draft tokens (90.58%).

### OCRBench V1 accuracy

A complete 1,000-sample OCRBench V1 evaluation completed successfully with no request errors and scored **826/1000 (82.6)**.

| Category | Score | Maximum |
| --- | ---: | ---: |
| Regular Text Recognition | 47 | 50 |
| Irregular Text Recognition | 44 | 50 |
| Artistic Text Recognition | 47 | 50 |
| Handwriting Recognition | 31 | 50 |
| Digit String Recognition | 35 | 50 |
| Non-Semantic Text Recognition | 46 | 50 |
| Scene Text-centric VQA | 178 | 200 |
| Doc-oriented VQA | 176 | 200 |
| Key Information Extraction | **171** | 200 |
| Handwritten Mathematical Expression Recognition | **51** | 100 |
| **Total** | **826** | **1000** |

Evaluation configuration: concurrency 64, temperature 1.0, top_p 0.95, max_tokens 8192, seed 7, thinking enabled, reasoning_effort=high.
For TP4 graph mode, capture batch sizes must be multiples of TP size. The
validated graph launch therefore uses `--max-num-seqs 4`; `--max-num-seqs 1`
is valid for eager smoke but not for this graph topology.

### User-facing change

The model architecture can now be loaded and served on Ascend using the
vLLM-compatible multimodal API. The validated W8A8 launch topology is TP4 /
DP4 / EP16 in eager or ACLGraph mode.

This PR is ready for implementation review. Marking it Ready for Review does
not yet claim full production qualification or inclusion in the supported
model matrix.

### Remaining release validation

- processor, ViT/aligner, router, hidden-state, and logits numerical comparison
  against the upstream NVIDIA/HF reference;
- full BF16 checkpoint validation and BF16/W8A8 accuracy comparison;
- maximum 384-token image spans, non-aligned boundaries, heterogeneous images,
  mixed batches, and longer-context cases;
- deployment alternatives such as TP8 / DP2;
- fixed image-task accuracy suite and text-only regression suite;
- startup, HBM, TTFT, TPOT, throughput, and regression measurements;
- supported-model documentation and a public launch example.
- automatic prefix caching passes with the DeepSeek-V4-specific contract:
  launch with `--block-size 32` and use prompts longer than 4K tokens. Three
  identical 5,014-token requests pinned to one DP rank reused 8,192 of 15,042
  prompt tokens; DSpark remained active with 15/18 accepted/draft tokens.
- the multimodal APC stack also passes: three identical 6,139-token
  long-prefix plus image requests all returned exact OCR text, reused 8,192
  prompt tokens, recorded two multimodal-processor cache hits, and kept DSpark
  active at 15/18 accepted/draft tokens.
- multimodal SSE streaming on the graph service reconstructs exact OCR text,
  reports `finish_reason=stop`, terminates with `[DONE]`, and exercises DSpark
  with 5/6 accepted/draft tokens.

First-phase exclusions remain DSA context parallel and FlashComm1. DSpark is
validated only as target-side multimodal prefill followed by text-only
speculative decode; the drafter does not consume image embeddings or propose
inside image spans.

Signed-off-by: GDzhu01 <116337067+GDzhu01@users.noreply.github.com>

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/models/deepseek_v4/model.py [hunk 1]
```diff
@@ -345,6 +354,8 @@ def __init__(
             routed_scaling_factor=self.routed_scaling_factor,
             swiglu_limit=self.swiglu_limit,
             e_score_correction_bias=self.gate.e_score_correction_bias,
+            bias_vl=self.gate.bias_vl,
+            image_sentinel_lo=129257,
```
Comment (dragondream-chen) [inline vllm_ascend/models/deepseek_v4/model.py]: Do magic numbers need to be annotated for explanation?

**Review by weijinqian0 — DISMISSED** [review]

**Comment by GDzhu01** [comment]
/cancel
[Bot]: successfully force-cancelled the following runs:


https://github.com/vllm-project/vllm-ascend/actions/runs/33831336723

**Review by weijinqian0 — APPROVED** [review]
